runner/src/examples.rs

```diff
@@ -0,0 +1,970 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{files::*, internal::*};
+use maplit::hashmap;
+use std::collections::HashMap;
+
+#[cfg(target_os = "macos")]
+const DEFAULT_SERVER_RUST_TARGET: &str = "x86_64-apple-darwin";
+#[cfg(not(target_os = "macos"))]
+const DEFAULT_SERVER_RUST_TARGET: &str = "x86_64-unknown-linux-musl";
+
+#[cfg(target_os = "macos")]
+const DEFAULT_EXAMPLE_BACKEND_RUST_TARGET: &str = "x86_64-apple-darwin";
+#[cfg(not(target_os = "macos"))]
+const DEFAULT_EXAMPLE_BACKEND_RUST_TARGET: &str = "x86_64-unknown-linux-gnu";
+
+pub const ALL_CLIENTS: &str = "all";
+pub const NO_CLIENTS: &str = "none";
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct Example {
+    name: String,
+    #[serde(default)]
+    server: ExampleServer,
+    #[serde(default)]
+    backends: HashMap<String, Executable>,
+    applications: Applications,
+    clients: HashMap<String, Executable>,
+}
+
+/// A construct representing either an Oak Classic or an Oak Functions application.
+///
+/// The condition that only one of `classic` or `functions` should be non-empty is
+/// checked in each operation of this struct. If neither or both are empty, the
+/// operation panics with an error message.
+#[derive(Debug)]
+enum Applications {
+    Classic(HashMap<String, ApplicationClassic>),
+    Functions(HashMap<String, ApplicationFunctions>),
+}
```

Comment:
I think https://serde.rs/enum-representations.html#internally-tagged may help here, have you looked into it?

Also, I think we can simplify this by pushing the enum one level deep:

```rust
#[serde(tag = "type")]
enum Application {
  Classic { ... }
  Functions { ... }
}

... applications: HashMap<String, Application>
```

I don't think we need to check that applications are all of the same type anyways, I mean it will be pretty obvious if they are mixed and don't work well together, and if and when we do have the use case for applications of different types in the same example, we would just remove that check anyways as the first thing.

WDYT?

Reply:
Thanks. I pushed an update, but I am not entirely sure what should be done in the following situations: 
- in case of an example that has an empty applications map (such as the `tls_attestation` example). I think my current implementation excludes it from both `run-examples` and `run-functions-examples`
- in case of an example that has a mix of both application types. We don't currently have such an example, but the current implementation could run both commands for it, in each case running only the applications of the relevant type.

Reply:
Changed it, so now if an example has a mix of both types of applications, it won't run.